nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py

```diff
@@ -0,0 +1,245 @@
+        hidden_states: torch.Tensor,
+        attention_mask: torch.Tensor,
+        encoder_output: Optional[torch.Tensor] = None,
+        enc_dec_attn_mask: Optional[torch.Tensor] = None,
+        inference_params: Optional[Any] = None,
+        is_first_microbatch: Optional[bool] = None,
+        checkpoint_core_attention: Optional[bool] = False,
+    ) -> torch.Tensor:
+        if self.dtype == torch.float32:
+            return super().forward(
+                hidden_states,
+                attention_mask,
+                encoder_output=encoder_output,
+                enc_dec_attn_mask=enc_dec_attn_mask,
+                inference_params=inference_params,
+                is_first_microbatch=is_first_microbatch,
+                checkpoint_core_attention=checkpoint_core_attention,
+            )
+        with torch.autocast(device_type="cuda", dtype=self.dtype):
+            return super().forward(
+                hidden_states,
+                attention_mask,
+                encoder_output=encoder_output,
+                enc_dec_attn_mask=enc_dec_attn_mask,
+                inference_params=inference_params,
+                is_first_microbatch=is_first_microbatch,
+                checkpoint_core_attention=checkpoint_core_attention,
+            )
+
+
+from megatron.core.transformer.transformer_layer import BaseTransformerLayer
+
+
+class TETransformerLayerAutocast(AutocastTransformerLayer, BaseTransformerLayer):
+    def __init__(self, config, layer_number=1, hidden_dropout=None):
+        self.config = config
+        self.is_first_microbatch = True
+        precision = 'bf16' if config.bf16 else 16
+
+        super().__init__(
+            hidden_size=config.hidden_size,
+            ffn_hidden_size=config.ffn_hidden_size,
+            layernorm_epsilon=config.layernorm_epsilon,
+            num_attention_heads=config.num_attention_heads,
+            init_method=config.init_method,
+            output_layer_init_method=config.output_layer_init_method,
+            hidden_dropout=config.hidden_dropout,
+            attention_dropout=config.attention_dropout,
+            layer_number=layer_number + self._get_layer_offset(),
+            kv_channels=config.kv_channels,
+            # self_attn_mask_type='causal', # Use default 'causal'
+            tp_size=parallel_state.get_tensor_model_parallel_world_size(),
+            params_dtype=config.params_dtype,
+            get_rng_state_tracker=tensor_parallel.random.get_cuda_rng_tracker,
+            fuse_wgrad_accumulation=config.gradient_accumulation_fusion,
+            seq_length=None,  # used for jit warmup
+            micro_batch_size=None,  # used for jit warmup
+            sequence_parallel=config.sequence_parallel,
+            apply_residual_connection_post_layernorm=config.apply_residual_connection_post_layernorm,
+            autocast_dtype=precision,
+            # use_emha=False, # Use default 'False'
+            ub_tp_comm_overlap=config.tp_comm_overlap,
```

Comment:
Still needs the TP knob changes from [here](https://github.com/NVIDIA/NeMo/pull/8290/files), in case you missed it